File: CS/15_Document/Template.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Spire.Pdf;
using Spire.Pdf.Graphics;
using Spire.Pdf.AutomaticFields;
using System.IO;

namespace Template
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            //Create a pdf document
            PdfDocument doc = new PdfDocument();
            doc.ViewerPreferences.PageLayout = PdfPageLayout.TwoColumnLeft;

            //Set the margin
            PdfUnitConvertor unitCvtr = new PdfUnitConvertor();
            PdfMargins margin = new PdfMargins();
            margin.Top = unitCvtr.ConvertUnits(2.54f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Bottom = margin.Top;
            margin.Left = unitCvtr.ConvertUnits(3.17f, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point);
            margin.Right = margin.Left;

            SetDocumentTemplate(doc, PdfPageSize.A4, margin);

            //Create one section
            PdfSection section = doc.Sections.Add();
            section.PageSettings.Size = PdfPageSize.A4;
            section.PageSettings.Margins = new PdfMargins(0);
            SetSectionTemplate(section, PdfPageSize.A4, margin, "Section 1");

            //Create one page
            PdfNewPage page = section.Pages.Add();

            //Draw page
            DrawPage(page); 

            page = section.Pages.Add();
            DrawPage(page);

            page = section.Pages.Add();
            DrawPage(page);

            page = section.Pages.Add();
            DrawPage(page);

            page = section.Pages.Add();
            DrawPage(page);

            //Save the document
            doc.SaveToFile("Template.pdf");
            doc.Close();

            //Launch the Pdf file
            PDFDocumentViewer("Template.pdf");
        }

        private void SetSectionTemplate(PdfSection section, SizeF pageSize, PdfMargins margin, String label)
        {
            // Create an element for the left blank space with the width of the left margin and the height of the page
            PdfPageTemplateElement leftSpace = new PdfPageTemplateElement(margin.Left, pageSize.Height);

            // Set the element as a foreground element
            leftSpace.Foreground = true;

            // Set the element as the left template for odd pages
            section.Template.OddLeft = leftSpace;

            // Create an Arial font object with a size of 9f and an italic style
            PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Arial", 9f, FontStyle.Italic));

            // Create a string format object with centered text alignment and vertically centered alignment
            PdfStringFormat format = new PdfStringFormat(PdfTextAlignment.Center, PdfVerticalAlignment.Middle);

            // Calculate the y-coordinate value to center the text on the page
            float y = (pageSize.Height - margin.Top - margin.Bottom) * (1 - 0.618f);

            // Create a rectangle object with the top-left corner at (10, y), and the bottom-right corner at (left margin - 20, font height + 6)
            RectangleF bounds = new RectangleF(10, y, margin.Left - 20, font.Height + 6);

            // Draw an orange-red rectangle within the left blank space element
            leftSpace.Graphics.DrawRectangle(PdfBrushes.OrangeRed, bounds);

            // Draw the label text within the rectangle using the Arial font created earlier, white color, and the previously created string format object
            leftSpace.Graphics.DrawString(label, font, PdfBrushes.White, bounds, format);

            // Create an element for the right blank space with the width of the right margin and the height of the page
            PdfPageTemplateElement rightSpace = new PdfPageTemplateElement(margin.Right, pageSize.Height);

            // Set the element as a foreground element
            rightSpace.Foreground = true;

            // Set the element as the right template for even pages
            section.Template.EvenRight = rightSpace;

            // Recalculate the bottom-right corner of the rectangle
            bounds = new RectangleF(10, y, margin.Right - 20, font.Height + 6);

            // Draw a brown rectangle within the right blank space element
            rightSpace.Graphics.DrawRectangle(PdfBrushes.SaddleBrown, bounds);

            // Draw the label text within the rectangle using the Arial font created earlier, white color, and the previously created string format object
            rightSpace.Graphics.DrawString(label, font, PdfBrushes.White, bounds, format);
        }


        // SetDocumentTemplate method for configuring document template with specified page size and margins
        private void SetDocumentTemplate(PdfDocument doc, SizeF pageSize, PdfMargins margin)
        {
            // Create a template element for the left space on the page
            PdfPageTemplateElement leftSpace = new PdfPageTemplateElement(margin.Left, pageSize.Height);
            doc.Template.Left = leftSpace;

            // Create a template element for the top space on the page
            PdfPageTemplateElement topSpace = new PdfPageTemplateElement(pageSize.Width, margin.Top);
            topSpace.Foreground = true;
            doc.Template.Top = topSpace;

            // Draw a header label
            PdfTrueTypeFont font = new PdfTrueTypeFont(new Font("Arial", 9f, FontStyle.Italic));
            PdfStringFormat format = new PdfStringFormat(PdfTextAlignment.Right);
            String label = "Demo of Spire.Pdf";
            SizeF size = font.MeasureString(label, format);
            float y = topSpace.Height - font.Height - 1;
            PdfPen pen = new PdfPen(Color.Black, 0.75f);
            topSpace.Graphics.SetTransparency(0.5f);
            topSpace.Graphics.DrawLine(pen, margin.Left, y, pageSize.Width - margin.Right, y);
            y = y - 1 - size.Height;
            topSpace.Graphics.DrawString(label, font, PdfBrushes.Black, pageSize.Width - margin.Right, y, format);

            // Create a template element for the right space on the page
            PdfPageTemplateElement rightSpace = new PdfPageTemplateElement(margin.Right, pageSize.Height);
            doc.Template.Right = rightSpace;

            // Create a template element for the bottom space on the page
            PdfPageTemplateElement bottomSpace = new PdfPageTemplateElement(pageSize.Width, margin.Bottom);
            bottomSpace.Foreground = true;
            doc.Template.Bottom = bottomSpace;

            // Draw a footer label
            y = font.Height + 1;
            bottomSpace.Graphics.SetTransparency(0.5f);
            bottomSpace.Graphics.DrawLine(pen, margin.Left, y, pageSize.Width - margin.Right, y);
            y = y + 1;
            PdfPageNumberField pageNumber = new PdfPageNumberField();
            PdfPageCountField pageCount = new PdfPageCountField();
            PdfCompositeField pageNumberLabel = new PdfCompositeField();
            pageNumberLabel.AutomaticFields = new PdfAutomaticField[] { pageNumber, pageCount };
            pageNumberLabel.Brush = PdfBrushes.Black;
            pageNumberLabel.Font = font;
            pageNumberLabel.StringFormat = format;
            pageNumberLabel.Text = "page {0} of {1}";
            pageNumberLabel.Draw(bottomSpace.Graphics, pageSize.Width - margin.Right, y);

            // Load and add a header image to the template as a stamp
            PdfImage headerImage = PdfImage.FromFile(@"..\..\..\..\..\..\Data\Header.png");
            PointF pageLeftTop = new PointF(-margin.Left, -margin.Top);
            PdfPageTemplateElement header = new PdfPageTemplateElement(pageLeftTop, headerImage.PhysicalDimension);
            header.Foreground = false;
            header.Graphics.SetTransparency(0.5f);
            header.Graphics.DrawImage(headerImage, 0, 0);
            doc.Template.Stamps.Add(header);

            // Load and add a footer image to the template as a stamp
            PdfImage footerImage = PdfImage.FromFile(@"..\..\..\..\..\..\Data\Footer.png");
            y = pageSize.Height - footerImage.PhysicalDimension.Height;
            PointF footerLocation = new PointF(-margin.Left, y);
            PdfPageTemplateElement footer = new PdfPageTemplateElement(footerLocation, footerImage.PhysicalDimension);
            footer.Foreground = false;
            footer.Graphics.SetTransparency(0.5f);
            footer.Graphics.DrawImage(footerImage, 0, 0);
            doc.Template.Stamps.Add(footer);
        }

        private void DrawPage(PdfPageBase page)
        {
            // Get the width of the page
            float pageWidth = page.Canvas.ClientSize.Width;
            float y = 0;

            // Title
            y = y + 5;

            // Create a black brush for drawing text
            PdfBrush brush2 = new PdfSolidBrush(Color.Black);

            // Set the font to Arial, size 16, and bold style
            PdfTrueTypeFont font2 = new PdfTrueTypeFont(new Font("Arial", 16f, FontStyle.Bold));

            // Create a string format for center alignment
            PdfStringFormat format2 = new PdfStringFormat(PdfTextAlignment.Center);

            // Set character spacing to 1
            format2.CharacterSpacing = 1f;

            // Set the text content
            String text = "Summary of Science";

            // Draw the text on the page using the specified format
            page.Canvas.DrawString(text, font2, brush2, pageWidth / 2, y, format2);

            // Measure the size of the text
            SizeF size = font2.MeasureString(text, format2);

            // Update the y position for the next element
            y = y + size.Height + 6;

            // Icon
            // Load the image from file
            PdfImage image = PdfImage.FromFile(@"..\..\..\..\..\..\Data\Wikipedia_Science.png");

            // Draw the image on the page at the specified position
            page.Canvas.DrawImage(image, new PointF(pageWidth - image.PhysicalDimension.Width, y));

            // Calculate the left space for the icon
            float imageLeftSpace = pageWidth - image.PhysicalDimension.Width - 2;

            // Calculate the bottom position for the icon
            float imageBottom = image.PhysicalDimension.Height + y;

            // Reference content
            // Set the font to Arial, size 9
            PdfTrueTypeFont font3 = new PdfTrueTypeFont(new Font("Arial", 9f));

            // Create a string format with no paragraph indentation
            PdfStringFormat format3 = new PdfStringFormat();

            // Set paragraph indentation to twice the size of the font
            format3.ParagraphIndent = font3.Size * 2;

            // Enable trailing spaces measurement
            format3.MeasureTrailingSpaces = true;

            // Set line spacing to 1.5 times the size of the font
            format3.LineSpacing = font3.Size * 1.5f;

            // Set the text content for the first line
            String text1 = "(All text and picture from ";

            // Set the text content for the second line
            String text2 = "Wikipedia";

            // Set the text content for the third line
            String text3 = ", the free encyclopedia)";

            // Draw the first line of text on the page using the specified format
            page.Canvas.DrawString(text1, font3, brush2, 0, y, format3);

            // Measure the size of the first line of text
            size = font3.MeasureString(text1, format3);

            // Update the x position for the second line of text
            float x1 = size.Width;

            // Reset paragraph indentation to zero
            format3.ParagraphIndent = 0;

            // Set the font to Arial, size 9, and underlined style
            PdfTrueTypeFont font4 = new PdfTrueTypeFont(new Font("Arial", 9f, FontStyle.Underline));

            // Set the brush color to blue
            PdfBrush brush3 = PdfBrushes.Blue;

            // Draw the second line of text on the page using the specified format
            page.Canvas.DrawString(text2, font4, brush3, x1, y, format3);

            // Measure the size of the second line of text
            size = font4.MeasureString(text2, format3);

            // Update the x position for the third line of text
            x1 = x1 + size.Width;

            // Draw the third line of text on the page using the specified format
            page.Canvas.DrawString(text3, font3, brush2, x1, y, format3);

            // Update the y position for the next element
            y = y + size.Height;

            // Content
            // Read all text from a file
            PdfStringFormat format4 = new PdfStringFormat();
            text = System.IO.File.ReadAllText(@"..\..\..\..\..\..\Data\Summary_of_Science.txt");

            // Set the font to Arial, size 10
            PdfTrueTypeFont font5 = new PdfTrueTypeFont(new Font("Arial", 10f));

            // Set line spacing to 1.5 times the size of the font
            format4.LineSpacing = font5.Size * 1.5f;

            // Create a string layouter for measuring and arranging text
            PdfStringLayouter textLayouter = new PdfStringLayouter();

            // Calculate the height of the content block based on the image bottom position and y position
            float imageLeftBlockHeight = imageBottom - y;

            // Layout the text content using the specified format and size
            PdfStringLayoutResult result = textLayouter.Layout(text, font5, format4, new SizeF(imageLeftSpace, imageLeftBlockHeight));
            if (result.ActualSize.Height < imageBottom - y)
            {
                // If the content is too tall for the available space, increase the height by the line height
                imageLeftBlockHeight = imageLeftBlockHeight + result.LineHeight;
                result = textLayouter.Layout(text, font5, format4, new SizeF(imageLeftSpace, imageLeftBlockHeight));
            }

            // Iterate through each line of text in the layout result
            foreach (LineInfo line in result.Lines)
            {
                // Draw the text on the page using the specified format and y position
                page.Canvas.DrawString(line.Text, font5, brush2, 0, y, format4);

                // Update the y position for the next line of text
                y = y + result.LineHeight;
            }
            // Create a text widget for rendering the remaining text content
            PdfTextWidget textWidget = new PdfTextWidget(result.Remainder, font5, brush2);

            // Create a text layout for specifying how to arrange the text content
            PdfTextLayout textLayout = new PdfTextLayout();

            // Set the break type to fit the page width
            textLayout.Break = PdfLayoutBreakType.FitPage;

            // Set the layout type to paginate the text content
            textLayout.Layout = PdfLayoutType.Paginate;

            // Define the bounds of the text content area on the page
            RectangleF bounds = new RectangleF(new PointF(0, y), page.Canvas.ClientSize);

            // Set the string format for the text widget
            textWidget.StringFormat = format4;

            // Draw the text content on the page using the specified format and layout
            textWidget.Draw(page, bounds, textLayout);
        }
        private void PDFDocumentViewer(string fileName)
        {
            try
            {
                System.Diagnostics.Process.Start(fileName);
            }
            catch { }
        }

    }
}

File: VB/15_Document/Template.vb
Imports Spire.Pdf
Imports Spire.Pdf.Graphics
Imports Spire.Pdf.AutomaticFields
Imports System.IO

Namespace Template
    Partial Public Class Form1
        Inherits Form
        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
            ' Create a new PdfDocument object
            Dim doc As New PdfDocument()

            ' Set the page layout of the viewer preferences to TwoColumnLeft
            doc.ViewerPreferences.PageLayout = PdfPageLayout.TwoColumnLeft

            ' Create a new PdfUnitConvertor object for unit conversion
            Dim unitCvtr As New PdfUnitConvertor()

            ' Create a new PdfMargins object for setting margins
            Dim margin As New PdfMargins()

            ' Set the top margin using unit conversion from centimeter to point
            margin.Top = unitCvtr.ConvertUnits(2.54F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)

            ' Set the bottom margin equal to the top margin
            margin.Bottom = margin.Top

            ' Set the left margin using unit conversion from centimeter to point
            margin.Left = unitCvtr.ConvertUnits(3.17F, PdfGraphicsUnit.Centimeter, PdfGraphicsUnit.Point)

            ' Set the right margin equal to the left margin
            margin.Right = margin.Left

            ' Set the document template with specified page size and margins
            SetDocumentTemplate(doc, PdfPageSize.A4, margin)

            ' Add a new section to the document
            Dim section As PdfSection = doc.Sections.Add()

            ' Set the page settings of the section to A4 size and zero margins
            section.PageSettings.Size = PdfPageSize.A4
            section.PageSettings.Margins = New PdfMargins(0)

            ' Set the section template with specified page size, margins, and title
            SetSectionTemplate(section, PdfPageSize.A4, margin, "Section 1")

            ' Add multiple pages to the section and draw their content
            Dim page As PdfNewPage = section.Pages.Add()
            DrawPage(page)

            page = section.Pages.Add()
            DrawPage(page)

            page = section.Pages.Add()
            DrawPage(page)

            page = section.Pages.Add()
            DrawPage(page)

            page = section.Pages.Add()
            DrawPage(page)

            ' Save the document to a file
            doc.SaveToFile("Template.pdf")

            ' Close the PDF document
            doc.Close()

            ' Launch the Pdf file
            PDFDocumentViewer("Template.pdf")
        End Sub

        Private Sub SetSectionTemplate(ByVal section As PdfSection, ByVal pageSize As SizeF, ByVal margin As PdfMargins, ByVal label As String)
            ' Create a new PdfPageTemplateElement for the left space of the page
            Dim leftSpace As New PdfPageTemplateElement(margin.Left, pageSize.Height)
            leftSpace.Foreground = True
            section.Template.OddLeft = leftSpace

            ' Create a new PdfTrueTypeFont for the text font
            Dim font As New PdfTrueTypeFont(New Font("Arial", 9.0F, FontStyle.Italic))
            ' =============================================================================
            ' Use the following code for netstandard dlls
            ' =============================================================================
            'Dim font As New PdfTrueTypeFont("Arial", 9.0F, FontStyle.Italic, True)
            ' =============================================================================

            ' Create a new PdfStringFormat for text alignment and vertical alignment
            Dim format As New PdfStringFormat(PdfTextAlignment.Center, PdfVerticalAlignment.Middle)

            ' Calculate the y-coordinate for the text placement based on page size and margins
            Dim y As Single = (pageSize.Height - margin.Top - margin.Bottom) * (1 - 0.618F)

            ' Define the bounds of the rectangle for drawing on the left space
            Dim bounds As New RectangleF(10, y, margin.Left - 20, font.Height + 6)

            ' Draw a rectangle with OrangeRed color on the left space
            leftSpace.Graphics.DrawRectangle(PdfBrushes.OrangeRed, bounds)

            ' Draw the label text in white color within the rectangle
            leftSpace.Graphics.DrawString(label, font, PdfBrushes.White, bounds, format)

            ' Create a new PdfPageTemplateElement for the right space of the page
            Dim rightSpace As New PdfPageTemplateElement(margin.Right, pageSize.Height)
            rightSpace.Foreground = True
            section.Template.EvenRight = rightSpace

            ' Define the bounds of the rectangle for drawing on the right space
            bounds = New RectangleF(10, y, margin.Right - 20, font.Height + 6)

            ' Draw a rectangle with SaddleBrown color on the right space
            rightSpace.Graphics.DrawRectangle(PdfBrushes.SaddleBrown, bounds)

            ' Draw the label text in white color within the rectangle
            rightSpace.Graphics.DrawString(label, font, PdfBrushes.White, bounds, format)
        End Sub

        Private Sub SetDocumentTemplate(ByVal doc As PdfDocument, ByVal pageSize As SizeF, ByVal margin As PdfMargins)
            ' Create a new PdfPageTemplateElement for the left space of the page
            Dim leftSpace As New PdfPageTemplateElement(margin.Left, pageSize.Height)
            doc.Template.Left = leftSpace

            ' Create a new PdfPageTemplateElement for the top space of the page
            Dim topSpace As New PdfPageTemplateElement(pageSize.Width, margin.Top)
            topSpace.Foreground = True
            doc.Template.Top = topSpace

            ' Create a new PdfTrueTypeFont for the text font
            Dim font As New PdfTrueTypeFont(New Font("Arial", 9.0F, FontStyle.Italic))
            ' =============================================================================
            ' Use the following code for netstandard dlls
            ' =============================================================================
            'Dim font As New PdfTrueTypeFont("Arial", 9.0F, FontStyle.Italic, True)
            ' =============================================================================


            ' Create a new PdfStringFormat for text alignment
            Dim format As New PdfStringFormat(PdfTextAlignment.Right)

            ' Define the label and measure its size with the specified font and format
            Dim label As String = "Demo of Spire.Pdf"
            Dim size As SizeF = font.MeasureString(label, format)

            ' Calculate the y-coordinate for drawing on the top space
            Dim y As Single = topSpace.Height - font.Height - 1

            ' Create a new PdfPen for drawing lines
            Dim pen As New PdfPen(Color.Black, 0.75F)

            ' Set transparency for the top space graphics
            topSpace.Graphics.SetTransparency(0.5F)

            ' Draw a line on the top space using the pen
            topSpace.Graphics.DrawLine(pen, margin.Left, y, pageSize.Width - margin.Right, y)

            ' Adjust the y-coordinate for drawing the label below the line
            y = y - 1 - size.Height

            ' Draw the label text in black color on the top space
            topSpace.Graphics.DrawString(label, font, PdfBrushes.Black, pageSize.Width - margin.Right, y, format)

            ' Create a new PdfPageTemplateElement for the right space of the page
            Dim rightSpace As New PdfPageTemplateElement(margin.Right, pageSize.Height)
            doc.Template.Right = rightSpace

            ' Create a new PdfPageTemplateElement for the bottom space of the page
            Dim bottomSpace As New PdfPageTemplateElement(pageSize.Width, margin.Bottom)
            bottomSpace.Foreground = True
            doc.Template.Bottom = bottomSpace

            ' Adjust the y-coordinate for drawing lines on the bottom space
            y = font.Height + 1

            ' Set transparency for the bottom space graphics
            bottomSpace.Graphics.SetTransparency(0.5F)

            ' Draw a line on the bottom space using the pen
            bottomSpace.Graphics.DrawLine(pen, margin.Left, y, pageSize.Width - margin.Right, y)

            ' Increment the y-coordinate for drawing text below the line
            y = y + 1

            ' Create instances of PdfPageNumberField and PdfPageCountField
            Dim pageNumber As New PdfPageNumberField()
            Dim pageCount As New PdfPageCountField()

            ' Create a composite field for displaying page number and total pages
            Dim pageNumberLabel As New PdfCompositeField()
            pageNumberLabel.AutomaticFields = New PdfAutomaticField() {pageNumber, pageCount}
            pageNumberLabel.Brush = PdfBrushes.Black
            pageNumberLabel.Font = font
            pageNumberLabel.StringFormat = format
            pageNumberLabel.Text = "page {0} of {1}"

            ' Draw the page number label on the bottom space
            pageNumberLabel.Draw(bottomSpace.Graphics, pageSize.Width - margin.Right, y)

            ' Load header and footer images from file
            Dim headerImage As PdfImage = PdfImage.FromFile("..\..\..\..\..\..\Data\Header.png")
            Dim footerImage As PdfImage = PdfImage.FromFile("..\..\..\..\..\..\Data\Footer.png")

            ' Define the position of the header image
            Dim pageLeftTop As New PointF(-margin.Left, -margin.Top)
            Dim header As New PdfPageTemplateElement(pageLeftTop, headerImage.PhysicalDimension)
            header.Foreground = False
            header.Graphics.SetTransparency(0.5F)

            ' Draw the header image on the page template
            header.Graphics.DrawImage(headerImage, 0, 0)

            ' Add the header template to the document's stamps collection
            doc.Template.Stamps.Add(header)

            ' Calculate the y-coordinate for the footer image placement
            y = pageSize.Height - footerImage.PhysicalDimension.Height

            ' Define the position of the footer image
            Dim footerLocation As New PointF(-margin.Left, y)
            Dim footer As New PdfPageTemplateElement(footerLocation, footerImage.PhysicalDimension)
            footer.Foreground = False
            footer.Graphics.SetTransparency(0.5F)

            ' Draw the footer image on the page template
            footer.Graphics.DrawImage(footerImage, 0, 0)

            ' Add the footer template to the document's stamps collection
            doc.Template.Stamps.Add(footer)
        End Sub

        Private Sub DrawPage(ByVal page As PdfPageBase)
            ' Retrieve the width of the page
            Dim pageWidth As Single = page.Canvas.ClientSize.Width

            ' Initialize the y-coordinate
            Dim y As Single = 0

            ' Increase the y-coordinate by 5
            y = y + 5

            ' Create a brush for text color and a font for text styling
            Dim brush2 As PdfBrush = New PdfSolidBrush(Color.Black)
            Dim font2 As New PdfTrueTypeFont(New Font("Arial", 16.0F, FontStyle.Bold))
            ' =============================================================================
            ' Use the following code for netstandard dlls
            ' =============================================================================
            'Dim font2 As New PdfTrueTypeFont("Arial", 9.0F, FontStyle.Bold, True)
            ' =============================================================================

            ' Create a format for text alignment and spacing
            Dim format2 As New PdfStringFormat(PdfTextAlignment.Center)
            format2.CharacterSpacing = 1.0F

            ' Set the text to be displayed
            Dim text As String = "Summary of Science"

            ' Draw the text on the page using the specified font, brush, position, and format
            page.Canvas.DrawString(text, font2, brush2, pageWidth / 2, y, format2)

            ' Measure the size of the drawn text
            Dim size As SizeF = font2.MeasureString(text, format2)

            ' Increase the y-coordinate by the height of the drawn text plus some padding
            y = y + size.Height + 6

            ' Load an image from a file
            Dim image As PdfImage = PdfImage.FromFile("..\..\..\..\..\..\Data\Wikipedia_Science.png")

            ' Draw the image on the page at a specific position
            page.Canvas.DrawImage(image, New PointF(pageWidth - image.PhysicalDimension.Width, y))

            ' Calculate the left space for the image and the bottom position
            Dim imageLeftSpace As Single = pageWidth - image.PhysicalDimension.Width - 2
            Dim imageBottom As Single = image.PhysicalDimension.Height + y

            ' Display reference content
            Dim font3 As New PdfTrueTypeFont(New Font("Arial", 9.0F))
            ' =============================================================================
            ' Use the following code for netstandard dlls
            ' =============================================================================
            'Dim font3 As New PdfTrueTypeFont("Arial", 9.0F, FontStyle.Regular, True)
            ' =============================================================================
            Dim format3 As New PdfStringFormat()
            format3.ParagraphIndent = font3.Size * 2
            format3.MeasureTrailingSpaces = True
            format3.LineSpacing = font3.Size * 1.5F

            ' Draw the first part of the reference text
            Dim text1 As String = "(All text and picture from "
            page.Canvas.DrawString(text1, font3, brush2, 0, y, format3)

            ' Measure the size of the drawn text
            size = font3.MeasureString(text1, format3)

            ' Calculate the starting position for the next part of the reference text
            Dim x1 As Single = size.Width
            format3.ParagraphIndent = 0

            ' Create a font and a brush for the underlined text
            Dim font4 As New PdfTrueTypeFont(New Font("Arial", 9.0F, FontStyle.Underline))
            ' =============================================================================
            ' Use the following code for netstandard dlls
            ' =============================================================================
            'Dim font4 As New PdfTrueTypeFont("Arial", 9.0F, FontStyle.Underline, True)
            ' =============================================================================
            Dim brush3 As PdfBrush = PdfBrushes.Blue

            ' Draw the second part of the reference text with underline
            Dim text2 As String = "Wikipedia"
            page.Canvas.DrawString(text2, font4, brush3, x1, y, format3)

            ' Measure the size of the drawn text
            size = font4.MeasureString(text2, format3)

            ' Update the starting position for the next part of the reference text
            x1 = x1 + size.Width

            ' Draw the third part of the reference text
            Dim text3 As String = ", the free encyclopedia)"
            page.Canvas.DrawString(text3, font3, brush2, x1, y, format3)

            ' Increase the y-coordinate by the height of the drawn text
            y = y + size.Height

            ' Configure the format for the main content
            Dim format4 As New PdfStringFormat()
            text = File.ReadAllText("..\..\..\..\..\..\Data\Summary_of_Science.txt")
            Dim font5 As New PdfTrueTypeFont(New Font("Arial", 10.0F))
            ' =============================================================================
            ' Use the following code for netstandard dlls
            ' =============================================================================
            'Dim font5 As New PdfTrueTypeFont("Arial", 10.0F, FontStyle.Regular, True)
            ' =============================================================================
            format4.LineSpacing = font5.Size * 1.5F

            ' Create a layouter for arranging the text
            Dim textLayouter As New PdfStringLayouter()

            ' Calculate the available height for the main content block
            Dim imageLeftBlockHeight As Single = imageBottom - y

            ' Layout the text to fit within the available space
            Dim result As PdfStringLayoutResult = textLayouter.Layout(text, font5, format4, New SizeF(imageLeftSpace, imageLeftBlockHeight))

            ' Adjust the available height if the actual content height is smaller
            If result.ActualSize.Height < imageBottom - y Then
                imageLeftBlockHeight = imageLeftBlockHeight + result.LineHeight
                result = textLayouter.Layout(text, font5, format4, New SizeF(imageLeftSpace, imageLeftBlockHeight))
            End If

            ' Draw each line of the main content
            For Each line As LineInfo In result.Lines
                page.Canvas.DrawString(line.Text, font5, brush2, 0, y, format4)
                y = y + result.LineHeight
            Next line

            ' Create a new instance of PdfTextWidget with specified parameters
            Dim textWidget As New PdfTextWidget(result.Remainder, font5, brush2)

            ' Create a new instance of PdfTextLayout
            Dim textLayout As New PdfTextLayout()

            ' Set the break type of textLayout to FitPage
            textLayout.Break = PdfLayoutBreakType.FitPage

            ' Set the layout type of textLayout to Paginate
            textLayout.Layout = PdfLayoutType.Paginate

            ' Create a new instance of RectangleF with specified parameters
            Dim bounds As New RectangleF(New PointF(0, y), page.Canvas.ClientSize)

            ' Set the string format of textWidget to format4
            textWidget.StringFormat = format4

            ' Draw the textWidget on the page within the specified bounds using the given textLayout
            textWidget.Draw(page, bounds, textLayout)
        End Sub

        Private Sub PDFDocumentViewer(ByVal fileName As String)
            Try
                Process.Start(fileName)
            Catch
            End Try
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load

        End Sub

    End Class
End Namespace
